main/src/library/List.flix

```diff
@@ -0,0 +1,527 @@
+
+    ///
+    /// Helper function for `mapWithIndex`.
+    /// Returns the result of applying `f` to every element in `xs` along with that element's index + `i`.
+    /// That is, the result is of the form: `f(x1, i) :: f(x2, i+1) :: ...`.
+    /// mapWithIndex(f, xs) = mapWithIndexHelper(f, xs, 0).
+    ///
+    def mapWithIndexHelper[a,b](f: (a, Int32) -> b, xs: List[a], i: Int32): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => f(x, i) :: mapWithIndexHelper(f, rs, i+1)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs` and concatenating the results.
+    /// Alias for `flatMap`.
+    ///
+    def concatMap[a,b](f: a -> List[b], xs: List[a]): List[b] = flatMap(f, xs)
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs` and concatenating the results.
+    /// Aliased as concatMap.
+    ///
+    def flatMap[a,b](f: a -> List[b], xs: List[a]): List[b] = match xs with {
+        case Nil => Nil
+        case x :: rs => append(f(x), flatMap(f, rs))
+    }
+
+    ///
+    /// Returns the reverse of `xs`.
+    ///
+    def reverse[a](xs: List[a]): List[a] = reverseHelper(xs, Nil)
+
+    ///
+    /// Helper function for `reverse`.
+    /// Returns `acc` appended to the reverse of `xs`.
+    /// reverse(xs) = reverseHelper(xs, Nil).
+    ///
+    def reverseHelper[a](xs: List[a], acc: List[a]): List[a] = match xs with {
+        case Nil => acc
+        case x :: rs => reverseHelper(rs, x :: acc)
+    }
+
+    ///
+    /// Returns `xs` with its elements rotated `n` positions to the left.
+    /// That is, returns a new list where the first `n` elements in `xs` are the last `n` elements of the new list.
+    /// If `n < 0` or `n > length(xs)` the semantics are undefined.
+    ///
+    def rotateLeft[a](n: Int32, xs: List[a]): List[a] = append(suffix(length(xs)-n, xs), prefix(n, xs))
+
+    ///
+    /// Returns `xs` with its elements rotated `n` positions to the right.
+    /// That is, returns a new list where the last `n` elements in `xs` are the first `n` elements of the new list.
+    /// If `n < 0` or `n > length(xs)` the semantics are undefined.
+    ///
+    def rotateRight[a](n: Int32, xs: List[a]): List[a] = append(suffix(n, xs), prefix(length(xs)-n, xs))
+
+    ///
+    /// Returns `xs` with the element at position `p` replaced by `x`.
+    /// If `p < 0` or `p > length(xs)-1` the semantics are undefined.
+    ///
+    def replace[a](p: Int32, x: a, xs: List[a]): List[a] = match (p, xs) with {
+        case (0, r :: rs) => x :: rs
+        case (i, r :: rs) => r :: replace(i-1, x, rs)
+    }
+
+    ///
+    /// Returns `ys` with the `n` elements starting at position `p` replaced with elements from `xs`.
+    /// If `n = 0` then `ys` is returned.
+    /// Otherwise, if `n < 0` or `p < 0` or `p + n > length(xs)` the semantics are undefined.
+    ///
+    def patch[a](p: Int32, n: Int32, xs: List[a], ys: List[a]): List[a] = match (p, n, xs, ys) with {
```

Comment:
I want to think more about this function. Two questions for you:

(1) How do you "read" this function? For example, when reading `map(f, xs)` you say: "map (i.e. apply) the function `f` over the elements `xs`.
(2) Is it possible to make this function work when some of the arguments are "out-of-bounds" or does that not make sense?